packages/cli/src/modules/insights/insights.pre-init.ts

```diff
@@ -0,0 +1,18 @@
+import type { ModulePreInitContext } from '../modules.config';
+
+export function shouldLoadModule(ctx: ModulePreInitContext): boolean {
+	// Only main instance should collect insights
+	// Because main instances are informed of all finished workflow executions, whatever the mode
+	if (ctx.instance.instanceType !== 'main') {
+		return false;
+	}
+
+	// Disable insights for sqlite if pool size is not set
+	// This is because legacy sqlite does not support nested transactions needed for insights
+	// TODO: remove once benchmarks confirm this issue is solved with buffering / flushing mechanism
+	if (ctx.database.type === 'sqlite' && !ctx.database.sqlite.poolSize) {
+		return false;
+	}
+
+	return true;
```

Comment:
Should we be checking directly for leader here? iirc `worker` and `webhook` have no instance role set. So this could also be one expression: `return ctx.instance.isLeader && ctx.dbType === 'sqlite' && poolSize > 0`

Reply:
Problem is: is the instance is not yet leader but becomes it afterwards, then it will not start the background process because the code listening to leader change will not be loaded: https://github.com/n8n-io/n8n/pull/14243/files#diff-4f9361d0f6ad08d01013feda9e8b68de0cae710a3937b42c16ce972ac902f2a5R27
I feel like doing the listening on the modules loading side instead could work but would be expanding the responsabilities of the module loader a lot (IMO)